### PR TITLE
__slots__

### PR DESCRIPTION
**Pull request by crusaderky**
What changes:
- Most classes now define ``__slots__``
- removed ``_initialized`` property
- Enforced checks that all subclasses must also define ``__slots__``. For third-party subclasses, this is for now a DeprecationWarning and should be changed into a hard crash later on.
- 22% reduction in RAM usage
- 5% performance speedup for a DataArray method that performs a ``_to_temp_dataset`` roundtrip 

**DISCUSS:** support for third party subclasses is very poor at the moment (#1097). Should we skip the deprecation altogether?

Performance benchmark:
```python
import timeit
import psutil
import xarray

a = xarray.DataArray([1, 2], dims=['x'], coords={'x': [10, 20]})
RUNS = 10000
t = timeit.timeit("a.roll(x=1, roll_coords=True)", globals=globals(), number=RUNS)
print("{:.0f} us".format(t / RUNS * 1e6))

p = psutil.Process()
N = 100000
rss0 = p.memory_info().rss
x = [
    xarray.DataArray([1, 2], dims=['x'], coords={'x': [10, 20]})
    for _ in range(N)
]
rss1 = p.memory_info().rss
print("{:.0f} bytes".format((rss1 - rss0) / N))
```
Output:

| test          | env | master     | slots      |
|:-------------:|:---:|:----------:| ----------:|
| DataArray.roll | py35-min | 332 us     | 360 us     |
|  DataArray.roll | py37 | 354 us     | 337 us     |
| RAM usage of a DataArray | py35-min     | 2755 bytes | 2074 bytes |
| RAM usage of a DataArray | py37     | 1970 bytes | 1532 bytes |

The performance degradation on Python 3.5 is caused by the deprecation mechanism - see changes to common.py.

I honestly never realised that xarray objects are measured in kilobytes (vs. 32 bytes of underlying buffers!)

### CODE REVIEW

File: xarray/core/common.py [hunk 1]
```diff
@@ -180,7 +187,25 @@ class AttrAccessMixin:
...
+    __slots__ = ()
+
+    def __init_subclass__(cls):
+        """Verify that all subclasses explicitly define ``__slots__``. If they don't,
+        raise error in the core xarray module and a DeprecationWarning in third-party
+        extensions.
+        This check is only triggered in Python 3.6+.
+        """
+        if not hasattr(object.__new__(cls), "__dict__"):
+            cls.__setattr__ = cls._setattr_slots
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply xarray/core/common.py]: Post deprecation warning and after dropping support for py35, there will only be ``__setattr__``, statically implemented with what is now the body of ``_setattr_slots``. ``__init_subclass__`` will retain only the health check.

**Comment by crusaderky** [comment]
@max-sixty any package that for any reason expects xarray objects to have a ``__dict__`` will break overnight. Any package that subclasses from xarray will get a warning.

**Comment by shoyer** [comment]
We definitely do not guarantee implementation details like xarray objects
having a __dict__. A warning for subclasses is already going above and
beyond IMO.

File: xarray/core/indexing.py [hunk 2]
```diff
@@ -793,6 +816,9 @@ class IndexingSupport:  # could inherit from enum.Enum on Python 3
     # for backends that support full vectorized indexer.
     VECTORIZED = "VECTORIZED"
 
+    def __new__(cls, *args, **kwargs):
+        raise TypeError("Static class")
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply xarray/core/indexing.py]: Done

File: xarray/core/common.py [hunk 3]
```diff
@@ -205,20 +230,52 @@ def __getattr__(self, name: str) -> Any:
...
+        """
         object.__setattr__(self, name, value)
+        if name in self.__dict__:
+            # Custom, non-slotted attr, or improperly assigned variable?
+            warnings.warn(
+                "Setting attribute %r on a %r object. Explicitly define __slots__ "
+                "to suppress this warning for legitimate custom attributes and "
+                "raise an error when attempting variables assignments."
+                % (name, type(self).__name__),
+                DeprecationWarning,
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply xarray/core/common.py]: Done

**Comment by crusaderky** [comment]
@shoyer this is theoretically done, however the design I had to come up with to preserve the current statefulness of user-defined accessors displeases me and I'd rather clean it up. See #3268 for design decision.

**Comment by crusaderky** [comment]
Added automated unit tests for the automated check on ``__slots__`` for all objects inside the xarray library. I could not find a straightforward way to automatically test that objects defined outside of the library log FutureWarning instead - I tested it by hand.

**Comment by crusaderky** [comment]
```python

In [1]: class A(xarray.DataArray): 
   ...:     pass 
   ...:                                                                                                                                                                                                                                                        
miniconda3/envs/xarray-py37/bin/ipython:1: FutureWarning: xarray subclass A should explicitly define __slots__

In [2]: a = A()                                                                                                                                                                                                                                                

In [3]: a.x = 1                                                                                                                                                                                                                                                
miniconda3/envs/xarray-py37/bin/ipython:1: FutureWarning: Setting attribute 'x' on a 'A' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.

In [4]: a.x                                                                                                                                                                                                                                                   
Out[4]: 1

In [5]: a.__dict__                                                                                                                                                                                                                                             
Out[5]: {'x': 1}
```

File: xarray/core/dataset.py [hunk 4]
```diff
@@ -860,6 +875,7 @@ def _replace(  # type: ignore
         attrs: "Optional[OrderedDict]" = __default,
         indexes: "Optional[OrderedDict[Any, pd.Index]]" = __default,
         encoding: Optional[dict] = __default,
+        accessors: Optional[Dict[str, Any]] = None,
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply xarray/core/dataset.py]: Done

**Comment by crusaderky** [comment]
Ready for merge if there are no additional comments

**Comment by shoyer** [comment]
thanks @crusaderky !

**Comment by max-sixty** [comment]
👍  Thanks @crusaderky

**Comment by crusaderky** [comment]
👍

**Pull request merged**